accepted/future-releases/constructor-tearoffs/faq.md

```diff
@@ -0,0 +1,256 @@
+
+**A:** Yes! *Thank you* for noticing! And that is a problem. With Dart 2.0 we introduced generic function invocations, and had to decide how to parse `f(a<b,c>(d))`. The argument(s) to `f` could be either two comparison operator expressions or a single generic function invocation. We decided on always choosing the latter when `b` and `c` can be parsed as types and the `>` is followed by a `(`. (We have to decide while we parse the program, long before we can even begin to figure out what `b` and `c` actually refer to, so the choice is entirely grammar based.) We now have even more similar ambiguous cases. For example `f(a<b,c>-d)` is ambiguous because `-` can both be a prefix operator after a greater-than operator, or an infix operator after an explicit type-instantiation. Our choice is to be very restrictive in when we parse `expr <` as starting a type argument. We only do so when the following tokens *can* be parsed as a type argument list, and the only if the *next token* after the final `>` of the type arguments is one of:
+
+> `)`, `}`, `]`, `;`, `:`, `,`,`(`, `.`, `==`, or `!=`
+
+If the next token is *any other token*, then the `<` is parsed as a less-than operator.
+
+**Q:** Can I call a static method from `List` on `List<int>`.
+
+**A:** No. You can write `List.copyRange` but not `List<int>.copyRange`. This is a grammar based restriction. Even if you have a type alias like `typedef Stupid<X> = int;`, where `Stupid<void>` just means `int`, you can't do `Stupid<void>.parseInt`. The only thing you can access through an instantiated type literal is constructors. (So `Stupid<void>.fromEnvironment` is valid, just please don't do it.)
+
+**Q:** What if I want to call an extension method defined on `Type` on a `List<int>`?
+
+**A:** Then you need parentheses, just like now. Writing `List<int>.filled` *only works for constructors*. If you write `.something` after an explicitly instantiated type literal, it tries to do a constructor lookup. If you write `.something` after a raw type literal, it tries to do a static member or constructor lookup (that's what you can do now). If you want to call instance or extension members on `Type`, whether an extension or just `.toString()`, you need to write `(List<int>).toString()`.
+
+**Q:** If I can instantiate a generic function, and `List.filled` is a generic function tear-off, why is `List.filled<int>` invalid?
+
+**A:** Because we say so. We could make it mean `(List.filled)<int>`, but we prefer to reserve the syntax for if/when we add proper generic constructors. And you *can* write the parentheses, just please don't. Prefer `List<int>.filled` if that's what you mean&mdash;and if it's not what you mean, you probably can't yet do what you mean.
+
+**Q:** So where can I write an explicit type argument instantiation.
+
+**A:** In short, after an expression *e*, where the type arguments are not then followed by an argument list (then it's just a generic invocation, not an instantiation), and where *e*:
```

Comment:
This answer is bascially just a restatement of the cases we enumerate in the spec.  I wonder if it would be more helpful to people as an FAQ if we simplified it and restated it with examples, e.g.:

> **A:** After any expression.  The meaning depends on what kind of expression it follows.  For example:

* If the expression is a class, mixin, or type alias, it just supplies type parameters to the type, e.g. `Type t = List<int>;`  Or, if the type is followed by a constructor name, then it's a constructor tear-off, e.g. `List<int> Function(int, int) f = List<int>.filled;`
* If the expression is a generic function, it instantiates it, e.g. `Iterable<int> Function(int Function(double)) = {1.0}.map<int>;`
* If the expression is a callable object with a generic `call` method, it instantiates the `call` method, e.g. if `foo.call` is a generic method, then `var f = foo<int>y` is equivalent to `var f = foo.call<int>;`

(Yes, I realize this glosses over a lot of details, but that's what a FAQ should do)

Reply:
Very nice! Taken and rephrased slightly.